plugins/kubernetes-backend/src/kubernetes-auth-translator/AzureIdentityKubernetesAuthTranslator.ts

```diff
@@ -31,11 +44,46 @@ export class AzureIdentityKubernetesAuthTranslator
       clusterDetails,
     );
 
-    const credentials = new DefaultAzureCredential();
-
-    // TODO: can we cache this? It's inneficiant to get a new token every time
-    const accessToken = await credentials.getToken(aksScope);
-    clusterDetailsWithAuthToken.serviceAccountToken = accessToken.token;
+    clusterDetailsWithAuthToken.serviceAccountToken = await this.getToken();
     return clusterDetailsWithAuthToken;
   }
+
+  private async getToken(): Promise<string> {
+    if (this.isTokenValid()) {
+      return this.accessToken.token;
+    }
+
+    if (!this.newToken) {
+      this.newToken = this.fetchNewToken();
+    }
+
+    return this.newToken;
+  }
+
+  private async fetchNewToken(): Promise<string> {
+    try {
+      this.logger.info('Fetching new Azure token for AKS');
+
+      const newAccessToken = await this.tokenCredential.getToken(aksScope, {
+        requestOptions: { timeout: 10_000 }, // 10 seconds
+      });
+      if (!newAccessToken) {
+        throw new Error('AccessToken is null');
+      }
+
+      this.accessToken = newAccessToken;
+    } catch (err) {
+      this.logger.error('Unable to fetch Azure token', err);
+      // don't throw the error, so the existing token will be re-used until we're able to fetch a new token
```

Comment:
I think an exception to this might make sense is the initial fetch. If we can't fetch a token straight away we don't want to be returning the empty token, and tbh I think it should be fine to throw in that case too.

Another possible situation to throw is if the existing token has expired, could be worth throwing the token refresh error in that case as I think that's probably more informative than the authorization that you'll get otherwise.

Reply:
Good point! Done + Test